linked_list/linked_list.py

```diff
@@ -1,79 +1,180 @@
+    # Time Complexity: O(n)
+    # Space Complexity: O(1)
     def length(self):
-        pass
+        #if there's no head, the length is zero
+        if not self.head:
+            return 0
+        #otherwise, how could I get the length?
+        #by iterating
+        #starting at the head, count until there's no self.next?
+        counter = 0
+        current_node = self.head
+        while current_node: #while there is a next node,
+            counter += 1 #add one to counter
+            current_node = current_node.next #make the current node the next node
+        
+        return counter
+
 
     # method that returns the value at a given index in the linked list
     # index count starts at 0
     # returns None if there are fewer nodes in the linked list than the index value
-    # Time Complexity: ?
-    # Space Complexity: ?
+    # Time Complexity: O(n)
+    # Space Complexity: O(1)
     def get_at_index(self, index):
-        pass
+        if not self.head:
+            return None
+        
+        i = 0
+        current_node = self.head
+        while current_node: 
+            if i == index:
+                return current_node.value
+            current_node = current_node.next 
+            i += 1 
+
+        return None
 
     # method that returns the value of the last node in the linked list
     # returns None if the linked list is empty
-    # Time Complexity: ?
-    # Space Complexity: ?
+    # Time Complexity: O()
+    # Space Complexity: O(1)
     def get_last(self):
-        pass
+        #if my list is empty
+        if not self.head:
+            return None
+        
+        #if my list is one node 
+        if self.head == self.tail:
+            return self.head.value
+
+        #if my list is longer than just one node
+        return self.tail.value
 
+        
     # method that inserts a given value as a new last node in the linked list
-    # Time Complexity: ?
-    # Space Complexity: ?
+    # Time Complexity: O(1)
+    # Space Complexity: O(1)
     def add_last(self, value):
-        pass
+        if not self.head:
+            self.head = self.tail = Node(value)
+        else:
+            new_node = Node(value, None, self.tail)
+            self.tail.next = new_node
+            self.tail = new_node
 
     # method to return the max value in the linked list
     # returns the data value and not the node
     def find_max(self):
-        pass
+        if not self.head:
+            return None
+
+        max = self.head.value
+        current_node = self.head
+
+        while current_node:
+            if current_node.value > max:
+                max = current_node.value
+            else:
+                current_node = current_node.next
+        
+        return max
 
     # method to delete the first node found with specified value
-    # Time Complexity: ?
-    # Space Complexity: ?
+    # Time Complexity: O(n)
+    # Space Complexity: O(n)
     def delete(self, value):

```

Comment:
👀 Nice implementation, however you aren't creating any new data structures or creating a call stack here so your space complexity would be O(1)